compiler/rustc_data_structures/src/sorted_map/index_map.rs

```diff
@@ -89,35 +85,13 @@ impl<I: Idx, K: Ord, V> SortedIndexMultiMap<I, K, V> {
     ///
     /// If there are multiple items that are equivalent to `key`, they will be yielded in
     /// insertion order.
-    pub fn get_by_key_enumerated<Q>(&self, key: &Q) -> impl '_ + Iterator<Item = (I, &V)>
-    where
-        Q: Ord + ?Sized,
-        K: Borrow<Q>,
-    {
-        match self.binary_search_idx(key) {
-            Err(_) => self.idxs_to_items_enumerated(&[]),
-
-            Ok(idx) => {
-                let start = self.idx_sorted_by_item_key[..idx]
-                    .partition_point(|&i| self.items[i].0.borrow() != key);
-                let end = idx
-                    + self.idx_sorted_by_item_key[idx..]
-                        .partition_point(|&i| self.items[i].0.borrow() == key);
-                self.idxs_to_items_enumerated(&self.idx_sorted_by_item_key[start..end])
-            }
-        }
-    }
-
-    fn binary_search_idx<Q>(&self, key: &Q) -> Result<usize, usize>
-    where
-        Q: Ord + ?Sized,
-        K: Borrow<Q>,
-    {
-        self.idx_sorted_by_item_key.binary_search_by(|&idx| self.items[idx].0.borrow().cmp(key))
-    }
-
-    fn idxs_to_items_enumerated(&'a self, idxs: &'a [I]) -> impl 'a + Iterator<Item = (I, &'a V)> {
-        idxs.iter().map(move |&idx| (idx, &self.items[idx].1))
+    pub fn get_by_key_enumerated(&'a self, key: K) -> impl '_ + Iterator<Item = (I, &V)> {
+        let lower_bound =
+            self.idx_sorted_by_item_key.partition_point(|&i| self.items[i].0.borrow() < &key);
+        self.idx_sorted_by_item_key[lower_bound..]
+            .iter()
+            .take_while(move |&&i| self.items[i].0.borrow().eq(&key))
+            .map(move |&idx| (idx, &self.items[idx].1))
```

Comment:
You can combine these in a single [`map_while`](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.map_while).

Reply:
Hm, IIUC `map_while` does an operation like "do `take_while` while `map` returns `Some`" and this doesn't seem to be a case?

Reply:
Yeah you'd have to return a Some if you have an item, or a None if it's done. So then you can check and map at the same time.

E.g. something like

```rust
.take_while(move |&i| {
    let (k, v) = &self.items[i];
    (k == key).then_some((i, v))     // or Some((i, v)).filter(|_| k == key)
})
```

Then you only need to index into `self.items[i]` once.

(But I don't think it really matters after optimization, so pick whatever you find more readable.)

Reply:
I see, thanks for the example! I don't have a strong preference but it seems better than accessing by a tuple indexing. Applied your suggestion.